### PR TITLE
Feat(replace-all-methods): Adds replace_all_objects, replace_all_rules and replace_all_synonyms

### PR DESCRIPTION
**Pull request by nunomaduro**
| Q                 | A
| ----------------- | ----------
| Bug fix?          | no
| New feature?      | yes
| BC breaks?        | no     
| Need Doc update   | yes

## What was changed

This Pull Request adds the methods `replace_all_objects`, `replace_all_rules` and `replace_all_synonyms` into the `Index::class`.

### CODE REVIEW

File: lib/algolia/index.rb [hunk 1]
```diff
@@ -333,6 +333,76 @@ def save_objects!(objects, request_options = {})
...
+    #
+    # @param objects the array of objects to save, each object must contain an objectID attribute
+    # @param request_options contains extra parameters to send with your query
+    #
+    def replace_all_objects(objects, request_options = {})
+      safe = request_options[:safe] || request_options['safe'] || false
+      request_options.delete(:safe)
+      request_options.delete('safe')
+
+      tmp_index = @client.init_index(@name + '_tmp_' + rand(10000000).to_s)
```
Comment (nunomaduro) [inline lib/algolia/index.rb]:
```suggestion
"#{@name}_tmp_"
```

File: lib/algolia/index.rb [hunk 2]
```diff
@@ -333,6 +333,76 @@ def save_objects!(objects, request_options = {})
...
+
+      responses = []
+
+      scope = ['settings', 'synonyms', 'rules']
+      res = @client.copy_index(@name, tmp_index.name, scope, request_options)
+      responses << res
+
+      if safe
+        wait_task(res['taskID'], WAIT_TASK_DEFAULT_TIME_BEFORE_RETRY, request_options)
```
Comment (nunomaduro) [inline lib/algolia/index.rb]: wait_task(res['taskID'], WAIT_TASK_DEFAULT_TIME_BEFORE_RETRY, request_options) if safe

File: lib/algolia/index.rb [hunk 3]
```diff
@@ -333,6 +333,76 @@ def save_objects!(objects, request_options = {})
...
+          count = 0
+        end
+      end
+
+      if batch.any?
+        res = tmp_index.save_objects(batch, request_options)
+        responses << res
+      end
+
+      if safe
```
Comment (nunomaduro) [inline lib/algolia/index.rb]:
```suggestion
      if safe
        responses.each do |res|
          tmp_index.wait_task(res['taskID'], WAIT_TASK_DEFAULT_TIME_BEFORE_RETRY, request_options)
        end
      end
```

File: lib/algolia/index.rb [hunk 4]
```diff
@@ -333,6 +333,76 @@ def save_objects!(objects, request_options = {})
...
+    end
+
+    #
+    # Override the current objects by the given array of objects and wait end of indexing
+    #
+    # @param objects the array of objects to save, each object must contain an objectID attribute
+    # @param request_options contains extra parameters to send with your query
+    #
+    def replace_all_objects!(objects, request_options = {})
+      replace_all_objects(objects, {'safe' => true})
```
Comment (nunomaduro) [inline lib/algolia/index.rb]: Join.

Comment (nunomaduro) [inline lib/algolia/index.rb]:
```suggestion
      replace_all_objects(objects, request_options.merge(safe: true))
```

File: lib/algolia/index.rb (hunk 3 above)
Comment (julienbourdeau) [reply lib/algolia/index.rb]: Looks like you want to make this change before we merge

**Pull request merged**